src/compiler/transformers/es6.ts

```diff
@@ -206,6 +410,116 @@ namespace ts {
+            convertedLoopState.nonLocalJumps |= Jump.Return;
+            return createReturn(
+                createObjectLiteral(
+                    [
+                        createPropertyAssignment(
+                            createIdentifier("value"),
+                            node.expression
+                                ? visitNode(node.expression, visitor, isExpression)
+                                : createVoidZero()
+                        )
+                    ]
+                )
+            );
+        }
+
+        function visitThisKeyword(node: Node): Node {
+            Debug.assert(convertedLoopState !== undefined);
+
+            if (useCapturedThis) {
+                // if useCapturedThis is true then 'this' keyword is contained inside the arrow function. 
+                convertedLoopState.containsLexicalThis = true;
+                return node;
+            }
+            return convertedLoopState.thisName || (convertedLoopState.thisName = createUniqueName("this"));
+        }
+
+        function visitIdentifier(node: Identifier): Identifier {
+            if (!convertedLoopState) {
+                return node;
+            }
+            if (isGeneratedIdentifier(node)) {
+                return node;
+            }
+            if (node.text !== "arguments" && !resolver.isArgumentsLocalBinding(<Identifier>getOriginalNode(node))) {
+                return node;
+            }
+            return convertedLoopState.argumentsName || (convertedLoopState.argumentsName = createUniqueName("arguments"));
+        }
+
+        function visitBreakOrContinueStatement(node: BreakOrContinueStatement): Statement {
+            if (convertedLoopState) {
+                // check if we can emit break\continue as is
+                // it is possible if either
+                //   - break\continue is labeled and label is located inside the converted loop
+                //   - break\continue is non-labeled and located in non-converted loop\switch statement
+                const jump = node.kind === SyntaxKind.BreakStatement ? Jump.Break : Jump.Continue;
+                const canUseBreakOrContinue =
+                    (node.label && convertedLoopState.labels && convertedLoopState.labels[node.label.text]) ||
+                    (!node.label && (convertedLoopState.allowedNonLabeledJumps & jump));
+
+                if (!canUseBreakOrContinue) {
+                    let labelMarker: string;
+                    if (!node.label) {
+                        if (node.kind === SyntaxKind.BreakStatement) {
+                            convertedLoopState.nonLocalJumps |= Jump.Break;
+                            labelMarker = "break";
+                        }
+                        else {
+                            convertedLoopState.nonLocalJumps |= Jump.Continue;
+                            // note: return value is emitted only to simplify debugging, call to converted loop body does not do any dispatching on it.

```

Comment:
`break` and `continue` inside converted loops are generated as `return` with some optional marker value that denotes jump target. Given that non-labeled `continue` emitted as `return` already terminates current iteration we don't need to handle it any specially on the callsite - the only purpose why we emit `return "continue"` instead of just `return` is to make debugging simpler.